silo-to-silo/contracts/src/SiloToSilo.sol

```diff
@@ -0,0 +1,352 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.17;
+
+import {IERC20 as IERC20_NEAR} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {AuroraSdk, NEAR, PromiseCreateArgs, PromiseResultStatus, PromiseWithCallback} from "@auroraisnear/aurora-sdk/aurora-sdk/AuroraSdk.sol";
+import "@auroraisnear/aurora-sdk/aurora-sdk/Utils.sol";
+import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";
+import "@openzeppelin/contracts/utils/Base64.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import "./IEvmErc20.sol";
+
+struct TokenInfo {
+    string nearTokenAccountId;
+    bool isStorageRegistered;
+}
+
+contract SiloToSilo is Initializable, UUPSUpgradeable, AccessControlUpgradeable, PausableUpgradeable {
+    using AuroraSdk for NEAR;
+    using AuroraSdk for PromiseCreateArgs;
+    using AuroraSdk for PromiseWithCallback;
+
+    bytes32 public constant CALLBACK_ROLE = keccak256("CALLBACK_ROLE");
+    bytes32 public constant PAUSE_ADMIN_ROLE = keccak256("PAUSE_ADMIN_ROLE");
+    bytes32 public constant UNPAUSE_ADMIN_ROLE = keccak256("UNPAUSE_ADMIN_ROLE");
+
+    uint64 constant BASE_NEAR_GAS = 10_000_000_000_000;
+    uint64 constant WITHDRAW_NEAR_GAS = 50_000_000_000_000;
+    uint64 constant FT_TRANSFER_CALL_NEAR_GAS = 150_000_000_000_000;
+    uint128 constant ASCII_0 = 48;
+    uint128 constant ASCII_9 = 57;
+    uint128 constant ONE_YOCTO = 1;
+    uint128 constant NO_DEPOSIT = 0;
+
+    NEAR public near;
+    string public siloAccountId;
+    uint256 public ftTransferCallCounter;
+
+    // auroraErc20Token => TokenInfo { nearTokenAccountId, isStorageRegistered }
+    mapping(IEvmErc20 => TokenInfo) public registeredTokens;
+
+    // auroraErc20Token => (userAddressOnAurora => userBalance)
+    mapping(IEvmErc20 => mapping(address => uint128)) public balance;
+
+    // auroraErc20Token => (nearAccountId => isStorageRegistered)
+    mapping(IEvmErc20 => mapping(string => bool)) public registeredRecipients;
+
+    event TokenRegistered(IEvmErc20 token, string nearAccountId);
+    event TokenStorageRegistered(IEvmErc20 token, string nearAccountId);
+    event RecipientStorageRegistered(IEvmErc20 token, string recipientId);
+    event Withdraw(IEvmErc20 token, address recipient, uint128 transferedAmount);
+    event InitFtTransferCall(uint256 nonce, address indexed sender, IEvmErc20 indexed token, string indexed receiverId);
+    event FtTransferCall(
+        uint256 nonce,
+        address indexed sender,
+        IEvmErc20 indexed token,
+        string indexed receiverId,
```

Comment:
```suggestion
    event InitFtTransferCall(uint256 indexed nonce, address indexed sender, IEvmErc20 indexed token, string receiverId);
    event FtTransferCall(
        uint256 indexed nonce,
        address indexed sender,
        IEvmErc20 indexed token,
        string receiverId,
```
Recommend indexing the nonce instead of receiverId, it will be useful to link the callback to the call. Also indexed strings don't work well for filtering: https://ethereum.stackexchange.com/questions/6840/indexed-event-with-string-not-getting-logged/7170#7170

Reply:
82c51e23b2387c7329b8f81c57bd6a695c49caa0